### ISSUE TITLE
Webhook should immediately 

### ISSUE THREAD

**Issue by MikeStall** [Azure/azure-functions-eventing-extension#10]
Currently, the http handler blocks on dispatching to the user code. 
https://github.com/Azure/azure-functions-eventing-extension/blob/c6abcec4689a0d78044df181242238d342058123/src/EventGridExtension/EventGridExtensionConfig.cs#L86-L102 

It should return immediately, and queue up dispatch to happen in the background.  We may need some support from SDK to enable this properly.

**Comment by davidebbo** [comment]
Issue title doesn't sound quite right :)

**Comment by drub0y** [comment]
So, while this is actually pretty easy to do, the fall out from this is that you would lose [the durability that Event Grid provides by default](https://docs.microsoft.com/en-us/azure/event-grid/delivery-and-retry) and the onus would be on the extension first to ensure the message was actually "delivered" to the function (e.g. the process didn't terminate before the function was executed on another thread) which it cannot guarantee for many reasons, then that burden would next be shifted to the function itself.

IMHO this would not be a good default behavior if you're aiming for "good pit of success". That said, maybe this is a "mode" you'd want to opt-into for some reason (e.g. you don't care about durability because it's temperature data coming from an IoT device). That's a totally valid use case, but then that's something that could be configured explicitly via a property on the `EventGridTriggerAttribute`. For example: `[EventGridTrigger(Durable = false)]`

**Comment by MikeStall** [comment]
@drub0y  - we'd queue before returning success, and the SDK would make it easy for extensions to do the right thing here. So EventGrid would be responsible for dispatching the webhook to the EventGrid extension, which would queue a message and return. This preserves the durability guarantee.  This also allows longer running webhook handlers.

**Comment by drub0y** [comment]
@MikeStall Sorry, when you say "we'd queue before returning success"... queue to what exactly? Obviously whatever it is needs to be durable itself. So are you saying the extension SDK would offer an extension point for this with in the box solution for like Azure Storage Queues and/or Azure Service Bus? 

Would this conflict with the fact that Azure Event Grid itself is going to offer that capability? Specifically I've seen/heard the team talking about in the box support for pushing Event Grid events onto SB topics/queues. 'Cause if I buy into that, which I think I personally would in some scenarios, then I'd just go back to using a function triggered via on the SB trigger rather than an Event Grid trigger in those scenarios.

**Comment by MikeStall** [comment]
@drub0y  - Azure Functions / WebJobs SDK maintains an internal Azure queue for executing functions. For example, Blob Trigger uses this today.  This bug is about generalizing that mechanism to allow other extensions to do it. 

Some other things to consider:
1. Long running:   EventGrid has a 60 second timeout. Other webhooks are shorter (SendGrid was 10 sec). Consider the case where a user function takes 65 seconds.  Naively, the event grid dispatch would keep seeing a timeout, and keep firing the webhook, leading to many executions.  With the queuing approach here, EventGrid would fire the event once, webjobs would converter it to a durable Azure queue message, and then it'd be safe if it went over 60 sec.  

2. Load Balancing:  Say an event grid notification comes in with 10 events. Naively, that'd need to be handled by a single by webhook handler. By pushing the events as queue messages, they'd get automatically load balanced to across multiple nodes.

**Comment by drub0y** [comment]
Ah, I see now, thanks for the clarification. 👍 

On your other points:

 1. Totally understand what you're saying, it's a _very_ common problem with resilient, distributed systems, right? I guess there's part of me that's like "oh yeah cool that _totally_ helps pit of success, do that!", but then there's another part of me is like "I don't want that getting in the way of my raw performance and I am totally willing to take on the responsibility of being idempotent to repeated events, so I want to opt-out!".
 2. Yeah, that batching is an interesting one to work around. Again I would argue this from both sides: do what you're saying to enable better fan out, but let people who maybe care a little more about the raw perf opt-out of that batch splitting/fan out behavior because they are willing to be idempotent in their own implementations.

So, to summarize, I believe I understand what you're saying now, but I would request that there be an opt-out path in place day one for people who don't want that added overhead getting in the way and maybe that is actually done the way I suggested with a property on the trigger attribute.

**Comment by MikeStall** [comment]
@drub0y  - the opt out is a great idea.

**Comment by jeffhollan** [comment]
Was using the Event Grid binding this week - ended up opting for HttpTrigger so I could debug locally in VS Code (separate issue). Just echoing I think we should do this item and allow for queuing and immediate return.

**Comment by girishacharya** [comment]
hi - Are we saying that timeout for webhook is 60 secs? If yes - then are we expecting developers to do asynchronous operations? I mean what if the work that I want to do as part of the webhook take more than 60 secs? But if we make that operation asynchronous and the work I want do fails then how do we recover from that situation because we already responded event grid 200 OK. In that case - would we lose the event?

**Comment by mathewc** [comment]
Related to core SDK work item https://github.com/Azure/azure-functions-eventgrid-extension/issues/10. Some work has been done for this but is not yet finished.

**Comment by zmarty** [comment]
We (internal Microsoft team) would love to have this working, many of our functions do not finish within the alloted time.

**Comment by zmarty** [comment]
Just to give you some more information on this, we have an ongoing incident where we see a high rate of Event Grid failures. We think what is happening is that Event Grid delivers a message to our endpoint, which starts the Azure Function, but it does not complete in time and Event Grid gives up and tries again with the same event, further increasing load on our system.
